test/test_flash_attention_varlen_backward.py

```diff
@@ -0,0 +1,283 @@
+      softmax_lse,
+      indices_q,
+      indices_k,
+      (cu_seqlens_q, cu_seqlens_k),
+      (max_seqlen_in_batch_q, max_seqlen_in_batch_k),
+  )
+
+
+@pytest.fixture(autouse=True, scope="module")
+def setup_env():
+  orign_env = os.getenv('PJRT_ALLOCATOR_FRACTION')
+  os.environ['PJRT_ALLOCATOR_FRACTION'] = '0.5'
+  yield
+  if orign_env is None:
+    os.environ.pop('PJRT_ALLOCATOR_FRACTION', None)
+  else:
+    os.environ['PJRT_ALLOCATOR_FRACTION'] = orign_env
+
+
+@pytest.mark.parametrize("dtype", [torch.bfloat16])
+@pytest.mark.parametrize("mha_type", ["gqa"])
+@pytest.mark.parametrize("deterministic", [False])
+@pytest.mark.parametrize("alibi", [False])
+@pytest.mark.parametrize("local", [False])
+@pytest.mark.parametrize("causal", [True])
+@pytest.mark.parametrize("d", [8])
+@pytest.mark.parametrize(
+    "seqlen_q,seqlen_k",
+    [
+        (8, 8),
+        (128, 128),
+        (2048, 2048),
+    ],
+)
+@pytest.mark.parametrize("dropout_p", [0.0])
+def test_flash_attn_backward(seqlen_q, seqlen_k, d, dropout_p, causal, local,
+                             alibi, deterministic, mha_type, dtype):
+  if d % 8 != 0:
+    pytest.skip(reason="Expected head_size_og % 8 == 0 to be true")
+
+  device = "cuda"
+  # set seed
+  torch.random.manual_seed(101)
+  batch_size = 4
+  nheads = 9
+  nheads_k = nheads if mha_type == "mha" else (1 if mha_type == "mqa" else 3)
+
+  assert nheads % nheads_k == 0
+  window_size = (-1, -1) if not local else tuple(
+      torch.randint(0, seqlen_k, (2,)).tolist())
+  torch.cuda.synchronize()
+  q = torch.randn(batch_size, seqlen_q, nheads, d, device=device, dtype=dtype)
+  softmax_scale = q.shape[-1]**(-0.5)
+  k = torch.randn(batch_size, seqlen_k, nheads_k, d, device=device, dtype=dtype)
+  v = torch.randn(batch_size, seqlen_k, nheads_k, d, device=device, dtype=dtype)
+  o = torch.randn(batch_size, seqlen_q, nheads, d, device=device, dtype=dtype)
+  do = torch.randn(batch_size, seqlen_q, nheads, d, device=device, dtype=dtype)
+  softmax_lse = torch.randn(
+      batch_size, nheads, seqlen_q, device=device, dtype=torch.float32)
+  rng_state = torch.Tensor([0, 0]).to(torch.int64).to(device)
+  dq = torch.empty_like(q)
+  dk = torch.empty_like(k)
+  dv = torch.empty_like(v)
+
+  attention_mask = torch.zeros(
+      batch_size, seqlen_k, dtype=torch.int32).to(device)
+  k_lengths = torch.randint(low=2, high=seqlen_k, size=(batch_size,))
+  for i in range(batch_size):
+    k_len = k_lengths[i].item()
+    attention_mask[i, :k_len] = 1
+    q[i, k_len:, :, :] = 0
+    k[i, k_len:, :, :] = 0
+    v[i, k_len:, :, :] = 0
+    o[i, k_len:, :, :] = 0
+    do[i, k_len:, :, :] = 0
+    softmax_lse[i, :, :k_len] = 0
+  q.requires_grad = True
+  k.requires_grad = True
+  v.requires_grad = True
+  o.requires_grad = True
+  softmax_lse.requires_grad = True
+
+  q_cuda, k_cuda, v_cuda, o_cuda, do_cuda, dq_cuda, dk_cuda, dv_cuda, \
+  softmax_lse_cuda, indices_q, indices_k, cu_seq_lens, max_seq_lens = _upad_input(
```

Comment:
why not use the unpad_input of FA?

Reply:
Customs call FA need static shape of inputs, and unpad is performed inside the custom call.

Reply:
Okay, it looks like this `_upad` function and FA's `unpad` function are not the same.
BTW, it may be typo, it should be `_unpad` here.

Reply:
Done